draft-thomson-ppm-l1-bound-sum.md

```diff
@@ -111,51 +111,126 @@ The difference is that Prio3L1BoundSum involves validation of
+The complete circuit is specified below.
 
 ~~~ python
-def eval(
-        self,
-        meas: list[F],
-        joint_rand: list[F],
-        num_shares: int) -> list[F]:
-    weight = 0
-    for i in range(self.length):
-        weight += self.field.decode_from_bit_vector(
-            meas[i * self.bits : (i + 1) * self.bits]
+class L1BoundSum(Valid[list[int], list[int], F]):
+    EVAL_OUTPUT_LEN = 2
+    length: int
+    bits: int
+    chunk_length: int
+    field: type[F]
+
+    def __init__(self,
+                 field: type[F],
+                 length: int,
+                 bits: int,
+                 chunk_length: int):
+        """
+        Instantiate the `L1BoundSum` circuit for measurements with
+        `length` elements, each in the range `[0, 2^bits)`, and with
+        a maximum L1 norm of `2^bits - 1`.
+        """
+        self.field = field
+        self.length = length
+        self.bits = bits
+        self.chunk_length = chunk_length
+        self.GADGETS = [ParallelSum(Mul(), chunk_length)]
+        self.GADGET_CALLS = [
+            ((length + 1) * bits + chunk_length - 1) // chunk_length
+        ]
+        self.MEAS_LEN = (length + 1) * bits
+        self.OUTPUT_LEN = length
+        self.JOINT_RAND_LEN = self.GADGET_CALLS[0]
+
+    def eval(
+            self,
+            meas: list[F],
+            joint_rand: list[F],
+            num_shares: int) -> list[F]:
+        range_check = self.field(0)
+        shares_inv = self.field(num_shares).inv()
+        for i in range(self.GADGET_CALLS[0]):
+            r = joint_rand[i]
+            r_power = r
+            inputs: list[Optional[F]]
+            inputs = [None] * (2 * self.chunk_length)
+            for j in range(self.chunk_length):
+                index = i * self.chunk_length + j
+                if index < len(meas):
+                    meas_elem = meas[index]
+                else:
+                    meas_elem = self.field(0)
+
+                inputs[j * 2] = r_power * meas_elem
+                inputs[j * 2 + 1] = meas_elem - shares_inv
+
+                r_power *= r
+
+            range_check += self.GADGETS[0].eval(
+                self.field,
+                cast(list[F], inputs),
+            )
+
+        observed_weight = self.field(0)
+        for i in range(self.length):
+            observed_weight += self.field.decode_from_bit_vector(
+                meas[i * self.bits:(i + 1) * self.bits]
+            )
+        weight_position = self.length * self.bits
+        claimed_weight = self.field.decode_from_bit_vector(
+            meas[weight_position:weight_position + self.bits]
         )
-    weight_bits = self.field.encode_into_bit_vector(weight)
-
-    sum_vec = SumVec(self.field, self.length + 1,
-                     self.bits, chunk_length)
-    return sum_vec.eval(meas + weight_bits, joint_rand, num_shares)
+        weight_check = observed_weight - claimed_weight
+
+        return [range_check, weight_check]
```

Comment:
Is there any value in making this `r**2*range_check + r*weight_check` or something similar?  It would increase the consumption of joint randomness, but it seems like we could save a bunch of communication.

Reply:
Previous VDAF drafts did just that, in fact.

When we return a vector of length greater than one out of a validity circuit, Prio3 now uses query randomness to create a random linear combination, compressing the two field elements. Thus, the communication cost isn't any different, because there's still one field element for the (reduced) circuit output in the verifier. (The soundness error of this compression is also slightly better, since we're using two query randomness values instead of two powers of one joint randomness value, but that's infinitesimal when dealing with just two output values)